dvc/objects/status.py

```diff
@@ -0,0 +1,182 @@
+import logging
+from typing import TYPE_CHECKING, Dict, Iterable, NamedTuple, Optional, Set
+
+from dvc.hash_info import HashInfo
+
+from .tree import Tree
+
+if TYPE_CHECKING:
+    from .db.base import ObjectDB
+    from .db.index import ObjectDBIndexBase
+    from .file import HashFile
+
+logger = logging.getLogger(__name__)
+
+
+class StatusResult(NamedTuple):
+    exists: Set["HashFile"]
+    missing: Set["HashFile"]
+
+
+class CompareStatusResult(NamedTuple):
+    ok: Set["HashFile"]
+    missing: Set["HashFile"]
+    new: Set["HashFile"]
+    deleted: Set["HashFile"]
+
+
+def _indexed_dir_hashes(odb, index, hash_infos, name, cache_odb):
+    from . import load
+
+    # Validate our index by verifying all indexed .dir hashes
+    # still exist on the remote
+    dir_hashes = {hash_info.value for hash_info in hash_infos}
+    indexed_dirs = set(index.dir_hashes())
+    indexed_dir_exists = set()
+    if indexed_dirs:
+        indexed_dir_exists.update(odb.list_hashes_exists(indexed_dirs))
+        missing_dirs = indexed_dirs.difference(indexed_dir_exists)
+        if missing_dirs:
+            logger.debug(
+                "Remote cache missing indexed .dir hashes '%s', "
+                "clearing remote index",
+                ", ".join(missing_dirs),
+            )
+            index.clear()
+
+    # Check if non-indexed (new) dir hashes exist on remote
+    dir_exists = dir_hashes.intersection(indexed_dir_exists)
+    dir_exists.update(odb.list_hashes_exists(dir_hashes - dir_exists))
+
+    # If .dir hash exists in the ODB, assume directory contents
+    # also exists
+    for dir_hash in dir_exists:
+        tree = load(cache_odb, HashInfo(name, dir_hash))
+        file_hashes = [entry.hash_info.value for _, entry in tree]
+        if dir_hash not in index:
+            logger.debug(
+                "Indexing new .dir '%s' with '%s' nested files",
+                dir_hash,
+                len(file_hashes),
+            )
+            index.update([dir_hash], file_hashes)
+        yield from file_hashes
+        yield tree.hash_info.value
+
+
+def _status_staging(objs: Iterable["HashFile"]) -> "StatusResult":
+    exists: Set["HashFile"] = set()
+    for obj in objs:
+        if isinstance(obj, Tree):
+            exists.update(entry for _, entry in obj)
+        exists.add(obj)
+    return StatusResult(exists, set())
+
+
+def status(
+    odb: "ObjectDB",
+    objs: Iterable["HashFile"],
+    name: Optional[str] = None,
+    index: Optional["ObjectDBIndexBase"] = None,
+    cache_odb: Optional["ObjectDB"] = None,
```

Comment:
`odb` passed into status will not always be where we want to load dir trees from (we want to load from local cache and not stream from the remote when possible)

For now this is just a performance optimization to keep parity with existing behavior, but the need for this in general will be more obvious once the full oid migration is done